src/dbnode/storage/series/lookup/entry.go

```diff
@@ -231,23 +223,9 @@ func (entry *Entry) IfAlreadyIndexedMarkIndexSuccessAndFinalize(
 	return successAlready
 }
 
-// RemoveIndexedForBlockStarts removes the entry for the index for all blockStarts.
-func (entry *Entry) RemoveIndexedForBlockStarts(
-	blockStarts map[xtime.UnixNano]struct{},
-) index.RemoveIndexedForBlockStartsResult {
-	var result index.RemoveIndexedForBlockStartsResult
-	entry.reverseIndex.Lock()
-	for k, state := range entry.reverseIndex.states {
-		_, ok := blockStarts[k]
-		if ok && state.success {
-			delete(entry.reverseIndex.states, k)
-			result.IndexedBlockStartsRemoved++
-			continue
-		}
-		result.IndexedBlockStartsRemaining++
-	}
-	entry.reverseIndex.Unlock()
-	return result
+// IsEmpty returns true if the entry has no in-memory series data.
+func (entry *Entry) IsEmpty() bool {
+	return entry.Series.IsEmpty()
```

Comment:
Instead of just doing `entry.Series.IsEmpty()` we will need to keep the idea of `RelookupAndIncrementReaderWriterCount` IMO which would re-lookup the latest lookup.Entry in the shard map.

There's a few reasons for this and we saw real problems without relooking up the entry. I can explain it more in depth if required but the problem is that sometimes a lookup.Entry is created for a new series which is passed off to the indexing queue but then a race of two datapoints for a series that doesn't exist yet causes only a single lookup.Entry to exist in the final shard map and the other one becomes orphaned (which could look empty but the lookup.Entry that made it into the shard map will have a series that is not actually empty).

That's why the caller (from within `mutable_segments.go`) should perform perhaps something like the following:
```go
			isEmpty, ok := entry.RelookupAndReturnIsEmpty()
			if !ok {
				// Should not happen since shard will not expire until
				// no more block starts are indexed.
				// We do not GC this series if shard is missing since
				// we open up a race condition where the entry is not
				// in the shard yet and we GC it since we can't find it
				// due to an asynchronous insert.
				return true
			}

			return !isEmpty
```

Reply:
Also we should probably add a metric for if the edge case happens using the `instrument.EmitAndLogInvariantError(...)` so that this panics scenario tests if this edge case ever starts occurring.

Reply:
Interesting wasn't aware of this edge case but it makes sense. Will add back this safeguard.

Reply:
PR updated accordingly